opcua/server/history_sql.py

```diff
@@ -137,19 +140,180 @@ def read_node_history(self, node_id, start, end, nb_values):
+                               'Timestamp TIMESTAMP, '
+                               '{fd})'.format(tn=table, fd=fields))
+
+            except sqlite3.Error as e:
+                self.logger.info('Historizing SQL Table Creation Error for events from %s: %s', source_id, e)
+
+            self._conn.commit()
 
     def save_event(self, event):
-        raise NotImplementedError
+        with self._lock:
+            _c_sub = self._conn.cursor()
+
+            table = self._get_table_name(event.SourceNode)
+
+            fields, placeholders, evtup = self._format_event(event)
+
+            # insert the event into the database
+            try:
+                _c_sub.execute('INSERT INTO "{tn}" ("Id", "Timestamp", {fd}) VALUES (NULL, "{ts}", '
+                               '{pl})'.format(tn=table, fd=fields, ts=event.Time, pl=placeholders), evtup)
+            except sqlite3.Error as e:
+                self.logger.error('Historizing SQL Insert Error for events from %s: %s', event.SourceNode, e)
+
+            self._conn.commit()
+
+            # get this node's period from the period dict and calculate the limit
+            period = self._datachanges_period[event.SourceNode]
+
+            if period:
+                # after the insert, if a period was specified delete all records older than period
+                date_limit = datetime.now() - period
+
+                try:
+                    _c_sub.execute('DELETE FROM "{tn}" WHERE Time < ?'.format(tn=table),
+                                   (date_limit.isoformat(' '),))
+                except sqlite3.Error as e:
+                    self.logger.error('Historizing SQL Delete Old Data Error for events from %s: %s', event.SourceNode, e)
+
+                self._conn.commit()
+
+    def read_event_history(self, source_id, start, end, nb_values, evfilter):
+        with self._lock:
+
+            _c_read = self._conn.cursor()
+
+            order = "ASC"
+
+            if start is None or start == ua.DateTimeMinValue:
+                order = "DESC"
+                start = ua.DateTimeMinValue
+
+            if end is None or end == ua.DateTimeMinValue:
+                end = datetime.utcnow() + timedelta(days=1)
+
+            if start < end:
+                start_time = start.isoformat(' ')
+                end_time = end.isoformat(' ')
+            else:
+                order = "DESC"
+                start_time = end.isoformat(' ')
+                end_time = start.isoformat(' ')
 
-    def read_event_history(self, start, end, evfilter):
-        raise NotImplementedError
+            if nb_values:
+                limit = nb_values + 1  # add 1 to the number of values for retrieving a continuation point
+            else:
+                limit = -1  # in SQLite a LIMIT of -1 returns all results
+
+            table = self._get_table_name(source_id)
+            clauses = self._get_select_clauses(source_id, evfilter)
```

Comment:
By TypId do you mean the select clause EventType?

This is because a specific event type would be selected via the WhereClause. For this to work EventType field would need to be selected.